codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java

```diff
@@ -712,4 +714,54 @@ public static Optional<String> getDocumentationOrJavadoc(Shape shape) {
       .map(StringTrait::getValue)
       .or(() -> shape.getTrait(JavaDocTrait.class).map(StringTrait::getValue));
   }
+
+  public static List<Shape> getTopologicallyOrderedOrphanedShapesForService(
+    ServiceShape serviceShape,
+    Model model
+  ) {
+    // Copy-paste Smithy-Core's shape discovery mechanism:
+    // Walk the model starting from the serviceShape.
+    // This generates shapes that are "known" to Smithy-Core's `generateShapesInService`.
+    Set<Shape> nonOrphanedShapes = new Walker(model).walkShapes(serviceShape);
+
+    // orphanedShapes = (all shapes in model) - (non-orphaned shapes)
+    Set<Shape> orphanedShapes = model.shapes().collect(Collectors.toSet());
+    orphanedShapes.removeAll(nonOrphanedShapes);
+
+    // Copy-paste Smithy-Core's shape ordering mechanism for topological ordering
+    // (Python needs topological ordering to write shapes in order; other languages might not matter)
+    List<Shape> orderedShapes = new ArrayList();
+
+    TopologicalIndex topologicalIndex = TopologicalIndex.of(model);
+
+    // Get orphaned shapes under the following conditions:
+    // 1. In the same namespace as the service. (Should only generate shapes in this service.)
+    // 2. Not a member shape. (Member shapes don't have their own shapes generated for them.)
+    for (Shape shape : topologicalIndex.getOrderedShapes()) {
+      if (
+        orphanedShapes.contains(shape) &&
+        shape
+          .getId()
+          .getNamespace()
+          .equals(serviceShape.getId().getNamespace()) &&
+        !shape.isMemberShape()
+      ) {
+        orderedShapes.add(shape);
+      }
+    }
+    for (Shape shape : topologicalIndex.getRecursiveShapes()) {
```

Comment:
I wasn't 100% sure from the TopologicalIndex documentation, but is this guaranteed to be a superset of `getOrderedShapes()`?

Would be nice to have an explicit unit test just for `getTopologicallyOrderedOrphanedShapesForService`, especially since it can clarify what is or isn't a recursive shape.

Reply:
I suspect they're disjoint, otherwise some shapes would be generated multiple times.

I think a unit test might be more effort than it's worth --

Primarily, I'm inclined to trust whatever Smithy-Core is doing here, but verify it against a TestModel/real projects. I built this PR against the MPL and DBESDK, it picked up orphaned shapes, and I can run the code. I don't think the test you're proposing gets us as much ROI.

For a unit test, we'd have to create a Smithy model as a POJO (sort of like [this](https://github.com/smithy-lang/smithy-dafny/blob/main-1.x/codegen/smithy-dafny-codegen/src/test/java/software/amazon/polymorph/util/TestModel.java))

Let me know your thoughts.

Reply:
> I suspect they're disjoint, otherwise some shapes would be generated multiple times.

Good point, I think they must be.

> For a unit test, we'd have to create a Smithy model as a POJO (sort of like [this](https://github.com/smithy-lang/smithy-dafny/blob/main-1.x/codegen/smithy-dafny-codegen/src/test/java/software/amazon/polymorph/util/TestModel.java)), which seems more brittle than is worth doing.

You can include a *.smithy file as a resource and read it in pretty easily instead: https://github.com/smithy-lang/smithy-dafny/blob/main-1.x/codegen/smithy-dafny-codegen/src/test/java/software/amazon/polymorph/smithyjava/modeled/ModeledShapeValueTest.java#L77-L82

But I'll leave it to your judgement on whether you want to bother with that on this PR, I won't block on it.

Reply:
That's pretty nifty -- I'll plan to do this when I do the OrphanedShapes TestModel, but not now.